File: ai/utils/pii-redaction.ts
// PII Redaction Utility for AI Processing - Phase 1

import { PIIRedactionResult } from './ai-types';

// Regex patterns for common PII
const PII_PATTERNS = {
  // Email addresses
  email: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g,
  
  // Phone numbers (Indian and international formats)
  phone: /(?:\+91[-.\s]?)?(?:\d{2,4}[-.\s]?)?(?:\d{3,4}[-.\s]?\d{3,4}|\d{10})\b/g,
  
  // Indian PAN numbers
  pan: /[A-Z]{5}[0-9]{4}[A-Z]{1}/g,
  
  // Aadhaar numbers
  aadhaar: /\b\d{4}\s?\d{4}\s?\d{4}\b/g,
  
  // Credit card numbers (basic pattern)
  creditCard: /\b(?:\d{4}[-.\s]?){3}\d{4}\b/g,
  
  // Bank account numbers (8-18 digits)
  bankAccount: /\b\d{8,18}\b/g,
  
  // Names (common Indian names - basic patterns)
  // This is simplified - in production, use NER models
  commonNames: /\b(?:Mr\.?|Mrs\.?|Ms\.?|Dr\.?)\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b/g,
  
  // Addresses (simplified - looks for patterns like "123 Street Name")
  address: /\b\d+\s+[A-Za-z\s]+(?:Street|St|Road|Rd|Avenue|Ave|Lane|Ln|Drive|Dr|Colony|Nagar)\b/gi,
  
  // Amounts in INR
  amounts: /(?:Rs\.?|INR|₹)\s*[\d,]+(?:\.\d{2})?/gi,
};

export class PIIRedactor {
  private redactionCounter = 0;
  
  /**
   * Redact PII from text while preserving context for AI processing
   */
  redactPII(text: string, preserveAmounts = true): PIIRedactionResult {
    let redactedText = text;
    const redactionMap: Record<string, string> = {};
    let containsPII = false;
    
    // Reset counter for each new redaction
    this.redactionCounter = 0;
    
    // Process each PII type
    Object.entries(PII_PATTERNS).forEach(([type, pattern]) => {
      // Skip amounts if preserveAmounts is true (for dispute amount context)
      if (type === 'amounts' && preserveAmounts) return;
      
      redactedText = redactedText.replace(pattern, (match) => {
        containsPII = true;
        const placeholder = this.generatePlaceholder(type);
        redactionMap[placeholder] = match;
        return placeholder;
      });
    });
    
    return {
      redactedText,
      redactionMap,
      containsPII,
    };
  }
  
  /**
   * Restore original PII from redacted text
   */
  restorePII(redactedText: string, redactionMap: Record<string, string>): string {
    let restoredText = redactedText;
    
    Object.entries(redactionMap).forEach(([placeholder, original]) => {
      restoredText = restoredText.replace(new RegExp(placeholder, 'g'), original);
    });
    
    return restoredText;
  }
  
  /**
   * Generate a contextual placeholder for redacted PII
   */
  private generatePlaceholder(type: string): string {
    this.redactionCounter++;
    
    const placeholders: Record<string, string> = {
      email: `[EMAIL_${this.redactionCounter}]`,
      phone: `[PHONE_${this.redactionCounter}]`,
      pan: `[PAN_${this.redactionCounter}]`,
      aadhaar: `[AADHAAR_${this.redactionCounter}]`,
      creditCard: `[CARD_${this.redactionCounter}]`,
      bankAccount: `[ACCOUNT_${this.redactionCounter}]`,
      commonNames: `[PERSON_${this.redactionCounter}]`,
      address: `[ADDRESS_${this.redactionCounter}]`,
      amounts: `[AMOUNT_${this.redactionCounter}]`,
    };
    
    return placeholders[type] || `[${type.toUpperCase()}_${this.redactionCounter}]`;
  }
  
  /**
   * Validate that text is safe for AI processing (no PII leaked)
   */
  validateSafeText(text: string): { isSafe: boolean; issues: string[] } {
    const issues: string[] = [];
    
    // Check for common PII patterns that might have been missed
    Object.entries(PII_PATTERNS).forEach(([type, pattern]) => {
      const matches = text.match(pattern);
      if (matches && matches.length > 0) {
        issues.push(`Potential ${type} found: ${matches.length} instances`);
      }
    });
    
    return {
      isSafe: issues.length === 0,
      issues,
    };
  }
}

// Export singleton instance
export const piiRedactor = new PIIRedactor();

// Utility function for quick redaction
export function redactPIIFromText(text: string, preserveAmounts = true): PIIRedactionResult {
  return piiRedactor.redactPII(text, preserveAmounts);
}

// Validation helper
export function validateTextSafety(text: string): boolean {
  return piiRedactor.validateSafeText(text).isSafe;
}

File: AI-Python/utils/security/pii_redaction.py
"""
PII Redaction Utility
Handles redaction of personally identifiable information
"""

import re
from typing import Dict, List, Tuple
from loguru import logger

class PIIRedactor:
    """Redacts PII from text while preserving context"""
    
    def __init__(self):
        self.redaction_counter = 0
        self.patterns = {
            "email": re.compile(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'),
            "phone": re.compile(r'(?:\+91[-.\s]?)?(?:\d{2,4}[-.\s]?)?(?:\d{3,4}[-.\s]?\d{3,4}|\d{10})\b'),
            "pan": re.compile(r'[A-Z]{5}[0-9]{4}[A-Z]{1}'),
            "aadhaar": re.compile(r'\b\d{4}\s?\d{4}\s?\d{4}\b'),
            "credit_card": re.compile(r'\b(?:\d{4}[-.\s]?){3}\d{4}\b'),
            "bank_account": re.compile(r'\b\d{8,18}\b'),
            "common_names": re.compile(r'\b(?:Mr\.?|Mrs\.?|Ms\.?|Dr\.?)\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b'),
            "address": re.compile(r'\b\d+\s+[A-Za-z\s]+(?:Street|St|Road|Rd|Avenue|Ave|Lane|Ln|Drive|Dr|Colony|Nagar)\b', re.IGNORECASE),
            "amounts": re.compile(r'(?:Rs\.?|INR|₹)\s*[\d,]+(?:\.\d{2})?', re.IGNORECASE)
        }
    
    def redact_text(self, text: str, preserve_amounts: bool = True) -> str:
        """Redact PII from text"""
        self.redaction_counter = 0
        redacted_text = text
        
        for pii_type, pattern in self.patterns.items():
            if pii_type == "amounts" and preserve_amounts:
                continue
                
            redacted_text = pattern.sub(
                lambda match: self._generate_placeholder(pii_type),
                redacted_text
            )
        
        return redacted_text
    
    def _generate_placeholder(self, pii_type: str) -> str:
        """Generate contextual placeholder for redacted PII"""
        self.redaction_counter += 1
        
        placeholders = {
            "email": f"[EMAIL_{self.redaction_counter}]",
            "phone": f"[PHONE_{self.redaction_counter}]",
            "pan": f"[PAN_{self.redaction_counter}]",
            "aadhaar": f"[AADHAAR_{self.redaction_counter}]",
            "credit_card": f"[CARD_{self.redaction_counter}]",
            "bank_account": f"[ACCOUNT_{self.redaction_counter}]",
            "common_names": f"[PERSON_{self.redaction_counter}]",
            "address": f"[ADDRESS_{self.redaction_counter}]",
            "amounts": f"[AMOUNT_{self.redaction_counter}]"
        }
        
        return placeholders.get(pii_type, f"[{pii_type.upper()}_{self.redaction_counter}]")
    
    def validate_safe_text(self, text: str) -> Tuple[bool, List[str]]:
        """Validate that text is safe for AI processing"""
        issues = []
        
        for pii_type, pattern in self.patterns.items():
            matches = pattern.findall(text)
            if matches:
                issues.append(f"Potential {pii_type} found: {len(matches)} instances")
        
        return len(issues) == 0, issues
